dolby_vision/src/xml/parser.rs

```diff
@@ -558,30 +704,181 @@ impl CmXmlParser {
+        let huevec_text = node
+            .children()
+            .find(|e| e.has_tag_name("HueVectorField"))
+            .unwrap()
+            .text()
+            .unwrap();
+
+        let huevec: Vec<u8> = huevec_text
+            .split(self.separator)
+            .map(|v| {
+                min(
+                    255,
+                    ((v.parse::<f32>().unwrap() * 128.0) + 128.0).round() as u8,
+                )
+            })
+            .collect();
+
+        ensure!(
+            huevec.len() == 6,
+            "Invalid L8 HueVectorField: should be 6 values"
+        );
+
+        // Set variable length according to the metadata
+        // Only write trims which were modified
+        let length = if huevec.iter().any(|v| *v != 128) {
+            25
+        } else if satvec.iter().any(|v| *v != 128) {
+            19
+        } else if clip_trim != 2048 {
+            13
+        } else if target_mid_contrast != 2048 {
+            12
+        } else {
+            10
+        };
+
         Ok(ExtMetadataBlockLevel8 {
+            length,
             target_display_index: target_display.id.parse::<u8>()?,
             trim_slope,
             trim_offset,
             trim_power,
             trim_chroma_weight,
             trim_saturation_gain,
             ms_weight,
+            target_mid_contrast,
+            clip_trim,
+            saturation_vector_field0: satvec[0],
+            saturation_vector_field1: satvec[1],
+            saturation_vector_field2: satvec[2],
+            saturation_vector_field3: satvec[3],
+            saturation_vector_field4: satvec[4],
+            saturation_vector_field5: satvec[5],
+            hue_vector_field0: huevec[0],
+            hue_vector_field1: huevec[1],
+            hue_vector_field2: huevec[2],
+            hue_vector_field3: huevec[3],
+            hue_vector_field4: huevec[4],
+            hue_vector_field5: huevec[5],
         })
     }
 
-    pub fn parse_level9_trim(&self, node: &Node) -> Result<ExtMetadataBlockLevel9> {
-        let source_color_model = node
+    fn find_primary_index(&self, primaries: &[f64; 8], is_source: bool) -> Result<u8> {
+        let presets = if is_source {
+            level9::PREDEFINED_COLORSPACE_PRIMARIES
+        } else {
+            level10::PREDEFINED_REALDEVICE_PRIMARIES
+        };
+
+        let matching_primaries = presets.iter().enumerate().find(|(_, preset_primaries)| {
+            primaries
+                .iter()
+                .zip(preset_primaries.iter())
+                .all(|(a, b)| (*a - *b).abs() < f64::EPSILON)
+        });
+
+        // Exact match to preset primaries
+        let primary_index = if let Some((primary_index, _)) = matching_primaries {
+            if is_source {
+                primary_index
+            } else {
+                // FIXME: Why are the target primaries offset by the preset source primaries?
+                primary_index + level9::PREDEFINED_COLORSPACE_PRIMARIES.len()
```

Comment:
I still don't understand why this is.
Probably going to have to experiment with different primary values.

Reply:
Anyways, this is starting to get pretty big so I'd rather merge before fixing more.

Reply:
Sorry for late, L10 should only use the `PREDEFINED_COLORSPACE_PRIMARIES`, while the `PREDEFINED_REALDEVICE_PRIMARIES` part is only for L9 from my test. That means L9 may have a `target_primary_index: 18` but L10 doesn’t.

If a target display uses primaries from the real devices list, then verifier do not recognize a `target_primary_index` derived from that list as predefined, we should use 255 in that moment.

Reply:
I'll have to play around with both primary/source index values.
If there's an issue with the current implementation, can create a new issue.